src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs

```diff
@@ -195,19 +195,27 @@ internal JsonTypeInfo(Type type, JsonConverter converter, JsonSerializerOptions
         }
 
         private bool _isConfigured;
+        private object _configureLock = new object();
 
         internal void EnsureConfigured()
         {
             if (_isConfigured)
                 return;
 
-            Configure();
+            lock (_configureLock)
+            {
+                if (_isConfigured)
+                    return;
+
+                Configure();
 
-            _isConfigured = true;
+                _isConfigured = true;
+            }
         }
 
         internal virtual void Configure()
         {
+            Debug.Assert(Monitor.IsEntered(_configureLock), "Configure called directly, use EnsureConfigured which locks this method");
```

Comment:
Since this is virtual, might make sense to add the same to any overrides?

Reply:
they will call base.Configure anyway

Reply:
OK. That might change in future iterations, but I guess nothing prevents users from overriding the Configure method elsewhere either.

Reply:
Should the method be changed to `protected virtual` instead?